# Описание проекта

**Цель проекта**: провести анализ базы данных сервиса для чтения книг по подписке. 

**Задачи**: 
* Посчитать, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитать количество обзоров и среднюю оценку;
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
* Определить автора с самой высокой средней оценкой книг с 50 и более оценками;
* Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Описание данных:**

Таблица **books** - содержит данные о книгах:
* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

Таблица **authors** - содержит данные об авторах:
* author_id — идентификатор автора;
* author — имя автора.

Таблица **publishers** - содержит данные об издательствах:
* publisher_id — идентификатор издательства;
* publisher — название издательства;

Таблица **ratings** - содержит данные о пользовательских оценках книг:
* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

Таблица **reviews** - содержит данные о пользовательских обзорах:
* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

# Подключение к БД и просмотр таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db']) 

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# функция для выведения основной информации по таблице
def table_info(table_name): 
    table = pd.io.sql.read_sql('SELECT * FROM ' + table_name, con = engine)
    display(table.head())
    return table.info()

In [5]:
# просмотр основной информации
table_info('books')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [6]:
# просмотр основной информации
table_info('authors')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [7]:
# просмотр основной информации
table_info('publishers')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [8]:
# просмотр основной информации
table_info('ratings')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [9]:
# просмотр основной информации
table_info('reviews')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Данные загружаются корректно, пропусков в данных нет. 

# Выполнение запросов

## Сколько книг вышло после 1 января 2000 года?

In [10]:
# посчитайте, сколько книг вышло после 1 января 2000 года
query = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
'''

In [11]:
# функция для выполнения запроса
def run_query(query):
    return pd.io.sql.read_sql(query, con = engine)

In [12]:
# выполнение запроса
run_query(query)

count
0    819

После 1 января 2000 года (не включая) вышла 819 книг.

## Запрос 2
* Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [13]:
# для каждой книги посчитайте количество обзоров и среднюю оценку
query = '''
WITH 
rev AS 
(SELECT book_id, COUNT(reviews.review_id) AS reviews_count
FROM reviews
GROUP BY book_id),

rate AS 
(SELECT book_id, AVG(ratings.rating) AS avg_rating
FROM ratings
GROUP BY book_id),

a AS 
(SELECT 
b.book_id, 
b.title,
au.author
FROM books AS b LEFT OUTER JOIN authors AS au ON b.author_id=au.author_id)

SELECT a.title, a.author, rev.reviews_count, ROUND(rate.avg_rating, 2) AS avg_rating
FROM rev FULL OUTER JOIN rate ON rev.book_id=rate.book_id
RIGHT OUTER JOIN a ON rate.book_id=a.book_id
'''

In [14]:
# выполнение запроса
result = run_query(query)
result.head()

title  \
0                                       'Salem's Lot   
1                 1 000 Places to See Before You Die   
2  13 Little Blue Envelopes (Little Blue Envelope...   
3  1491: New Revelations of the Americas Before C...   
4                                               1776   

                           author  reviews_count  avg_rating  
0  Stephen King/Jerry N. Uelsmann            2.0        3.67  
1                Patricia Schultz            1.0        2.50  
2                 Maureen Johnson            3.0        4.67  
3                 Charles C. Mann            2.0        4.50  
4                David McCullough            4.0        4.00

In [15]:
# информация по результату
result.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   title          1000 non-null   object 
 1   author         1000 non-null   object 
 2   reviews_count  994 non-null    float64
 3   avg_rating     1000 non-null   float64
dtypes: float64(2), object(2)
memory usage: 31.4+ KB


Для всех книг определён рейтинг, но только для 994 книг есть обзоры.

## Запрос 3
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [16]:
# определите издательство, которое выпустило наибольшее число книг толще 50 страниц
query = '''
WITH b50 AS 
(SELECT book_id, publisher_id
FROM BOOKS
WHERE num_pages > 50)

SELECT 
p.publisher,
COUNT(book_id) AS books_count
FROM b50 LEFT OUTER JOIN publishers AS p ON b50.publisher_id=p.publisher_id
GROUP BY p.publisher_id
ORDER BY books_count DESC
LIMIT 1
'''

In [17]:
# выполнение запроса
run_query(query)

publisher  books_count
0  Penguin Books           42

Издательство "Penguin Books" выпустило больше всех книг толще 50 страниц - 42 книги.

## Запрос 4 
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками. 

In [18]:
# определите автора с самой высокой средней оценкой книг
query = '''
WITH b AS
(SELECT 
book_id, 
COUNT(rating) AS rating_count,
AVG(rating) AS avg_rating
FROM ratings
GROUP BY book_id)

SELECT a.author,
ROUND(AVG(avg_rating), 2) AS avg_rating
FROM
(SELECT book_id, avg_rating
FROM b 
WHERE rating_count >= 50) AS b50
LEFT OUTER JOIN 
(SELECT book_id, author_id
FROM books) AS ba ON b50.book_id=ba.book_id
LEFT OUTER JOIN 
authors AS a ON ba.author_id=a.author_id
GROUP BY a.author_id
ORDER BY avg_rating DESC
LIMIT 1
'''

In [19]:
# выполнение запроса
run_query(query)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.28

Средняя оценка книг для писателя считалась как средняя средних оценок книги. Самым высоким средним рейтингом книг обладает Джоан Роулинг (видимо для книг с иллюстрациями Мэри ГрандПри) - 4.28 из 5.

## Запрос 5 
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [20]:
# посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок
query = '''
WITH u AS 
(SELECT 
username,
COUNT(rating_id) AS rating_count
FROM ratings
GROUP BY username)

SELECT ROUND(AVG(reviews_count)) AS avg_reviews_count
FROM
(SELECT r.username,
COUNT(review_id) AS reviews_count
FROM
(SELECT username
FROM u
WHERE rating_count > 50) AS u50
LEFT OUTER JOIN reviews AS r ON u50.username=r.username
GROUP BY r.username) AS final
'''

In [21]:
# выполнение запроса
run_query(query)

avg_reviews_count
0               24.0

Получилось, что пользователи, которые выставляют много оценок (больше 50), в среднем написали 24 отзыва.

# Вывод

* В БД сожержится информация о 819 книгах, вышедших после 1 января 2000 года;
* Обзоры есть только у 994 книг из 1000;
* Больше всех книг толщиной больше 50 страниц напечатано издательством "Penguin Books";
* Автор с самой высокой средней оценкой книг — Джоан Роалинг;
* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3.